src/couch_views/src/couch_views_reduce_fdb.erl

```diff
@@ -0,0 +1,248 @@
+    } = TxDb,
+
+    lists:foreach(fun(View) ->
+        #mrview{
+            reduce_funs = ViewReduceFuns,
+            id_num = ViewId
+        } = View,
+
+        ExistingKeys = get_existing_keys(ViewId, ExistingViewKeys),
+        lists:foreach(fun(ViewReduceFun) ->
+            Tree = open_tree(TxDb, Sig, ViewId, ViewReduceFun),
+            delete_keys(Tx, Tree, DocId, ExistingKeys)
+
+        end, ViewReduceFuns)
+
+    end, Views);
+
+write_doc(TxDb, Sig, Views, #{reduce_results := ViewReduceResults, id := DocId},
+    ExistingViewKeys) ->
+    #{
+        tx := Tx
+    } = TxDb,
+
+    lists:foreach(fun({View, ReduceResults}) ->
+        #mrview{
+            reduce_funs = ViewReduceFuns,
+            id_num = ViewId
+        } = View,
+
+        ExistingKeys = get_existing_keys(ViewId, ExistingViewKeys),
+
+        lists:foreach(fun({ViewReduceFun, ReduceResult}) ->
+            Tree = open_tree(TxDb, Sig, ViewId, ViewReduceFun),
+
+            delete_keys(Tx, Tree, DocId, ExistingKeys),
+            add_keys(Tx, Tree, DocId, ReduceResult)
+
+        end, lists:zip(ViewReduceFuns, ReduceResults))
+    end, lists:zip(Views, ViewReduceResults));
+
+write_doc(_TxDb, _Sig, _Views, _Doc, _ExistingViewKeys) ->
+    ok.
+
+
+idx_prefix(DbPrefix, Sig, ViewId) ->
+    Key = {?DB_VIEWS, ?VIEW_DATA, Sig, ?VIEW_REDUCE_RANGE, ViewId},
+    erlfdb_tuple:pack(Key, DbPrefix).
+
+
+get_kv_size(TxDb, Mrst, ViewId) ->
+    #mrst {
+        views = Views,
+        sig = Sig
+    } = Mrst,
+
+    #{
+        tx := Tx
+    } = TxDb,
+
+    [View] = lists:filter(fun(View) -> View#mrview.id_num == ViewId end, Views),
+    #mrview{
+        reduce_funs = ViewReduceFuns,
+        id_num = ViewId
+    } = View,
+
+    lists:foldl(fun(ReduceFun, Acc) ->
+        Tree = open_tree(TxDb, Sig, ViewId, ReduceFun),
+        {_, KVSize} = ebtree:full_reduce(Tx, Tree),
+        Acc + KVSize
+    end, 0, ViewReduceFuns).
+
+
+fold(Db, Sig, ViewId, Reducer, GroupLevel, Opts, UserCallback,
+    UserAcc0) ->
+
+    Acc0 = #{
+        user_callback => UserCallback,
+        user_acc => UserAcc0,
+        reducer => Reducer
+    },
+
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        #{
+            tx := Tx
+        } = TxDb,
+
+        StartKey = fabric2_util:get_value(start_key, Opts),
+        EndKey = fabric2_util:get_value(end_key, Opts),
+
+        Tree = open_tree(TxDb, Sig, ViewId, Reducer),
+        GroupKeyFun = fun({Key, _DocId}) ->
+            couch_views_util:group_level_key(Key, GroupLevel)
+        end,
+
+        Acc1 = ebtree:group_reduce(Tx, Tree, StartKey, EndKey, GroupKeyFun,
+            fun fold_cb/2, Acc0, Opts),
+
+        maps:get(user_acc, Acc1)
+    end).
+
+
+get_existing_keys(ViewId, ExistingViewKeys) ->
+    case lists:keyfind(ViewId, 1, ExistingViewKeys) of
+        {ViewId, _TotalRows, _TotalSize, EKeys} ->
+            EKeys;
+        false ->
+            []
+    end.
+
+
+% The reduce values are stored as keys in the b-tree
```

Comment:
Can you clarify here? I _think_ the reduce value you've calculated outside of ebtree is the reduction over the k-v's emitted by a single document?